pilot/cmd/pilot-agent/readiness.go

```diff
@@ -0,0 +1,239 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+
+	"errors"
+
+	"time"
+
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	statsRequestTimeout     = time.Second * 2
+	readyStalenessThreshold = time.Minute * 5
+	statXdsActive           = "cluster.xds-grpc.upstream_cx_active"
+	statCdsUpdates          = "cluster_manager.cds.update_success"
+	statWarmingClusters     = "cluster_manager.warming_clusters"
+	statLdsUpdates          = "listener_manager.lds.update_success"
+	statWarmingListeners    = "listener_manager.total_listeners_warming"
+)
+
+type readinessProbe struct {
+	adminPort         int
+	lastReadySnapshot *statsSnapshot
+	lastErr           error
+	mutex             sync.Mutex
+}
+
+func newReadinessProbe(adminPort int) *readinessProbe {
+	return &readinessProbe{
+		adminPort: adminPort,
+	}
+}
+
+func (p *readinessProbe) checkReady() error {
+	// Get a new snapshot of the Envoy stats.
+	s, err := p.newStatsSnapshot()
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if err != nil {
+		// Failed getting the current snapshot. Envoy may be restarting. Reinitialize the ready probe.
+		p.lastReadySnapshot = nil
+		p.lastErr = nil
+		return err
+	}
+
+	// Case 1: Wait for the first ready snapshot
+	if p.lastReadySnapshot == nil {
+		if s.active && s.isReady() {
+			// Only save the first ready snapshot
+			return p.transitionToReady(s)
+		}
+		return p.transitionToNotReady(s)
+	}
+
+	// Case 2: the last result was not ready
+	if p.lastErr != nil {
+		if s.active {
+			if s.isReady() && s.isUpdatedSince(p.lastReadySnapshot) {
+				// Switch back to ready.
+				return p.transitionToReady(s)
+			}
+			// Discard this snapshot even though it's active, since
+			// it hasn't been updated since going active.
+		}
+
+		// Continue publishing ready up to the timeout
+		return p.transitionToNotReady(s)
+	}
+
+	// Case 3: last snapshot was active
+	if s.active {
+		// Update the last active snapshot
+		return p.transitionToReady(s)
+	}
+	// Continue publishing ready up to the timeout.
+	return p.transitionToNotReady(s)
+}
+
+func (p *readinessProbe) transitionToReady(s *statsSnapshot) error {
+	p.lastReadySnapshot = s
+	p.lastErr = nil
+	return nil
+}
+
+func (p *readinessProbe) transitionToNotReady(s *statsSnapshot) error {
+	if p.lastReadySnapshot == nil || s.t.After(p.lastReadySnapshot.t.Add(readyStalenessThreshold)) {
+		p.lastErr = errors.New(s.String())
+	} else {
+		// Continue publishing the previous ready state until the staleness timeout.
+		p.lastErr = nil
+	}
+	return p.lastErr
+}
+
+type statsSnapshot struct {
+	active           bool
```

Comment:
Nit: perhaps `connected` would be a better name? Though I'm not sure if that field is needed at all.

Reply:
Ack